File: 第七章 堆和优先队列/MaxHeap.java
// 使用数组
public class MaxHeap<E extends Comparable<E>> {

    private Array<E> data;

    public MaxHeap(int capacity) {
        data = new Array<>(capacity);
    }

    public MaxHeap() {
        data = new Array<>();
    }

    // 返回堆中的元素个数
    public int size() {
        return data.getSize();
    }

    // 返回一个布尔值, 表示堆中是否为空
    public boolean isEmpty() {
        return data.isEmpty();
    }

    // 返回完全二叉树的数组表示中，一个索引所表示的元素的父亲节点的索引
    private int parent(int index) {
        if (index == 0)
            throw new IllegalArgumentException("index-0 doesn't have parent.");
        return (index - 1) / 2;
    }

    // 返回完全二叉树的数组表示中，一个索引所表示的元素的左孩子节点的索引
    private int leftChild(int index) {
        return index * 2 + 1;
    }

    // 返回完全二叉树的数组表示中，一个索引所表示的元素的右孩子节点的索引
    private int rightChild(int index) {
        return index * 2 + 2;
    }

    public void add(E e) {
        data.addLast(e);
        shiftUp(data.getSize() - 1);
    }

    // shift up
    private void shiftUp(int k) {
        while (k > 0 && data.get((parent(k))).compareTo(data.get(k)) < 0) {
            data.swap(k, parent(k));
            k = parent(k);
        }
    }

    // 看堆中的最大元素
    public E findMax() {
        if (data.getSize() == 0)
            throw new IllegalArgumentException("Can not findMax when heap is empty.");
        return data.get(0);
    }

    // 取出堆中最大元素
    public E extractMax() {

        E ret = findMax();
        data.swap(0, data.getSize() - 1);
        data.removeLast();
        siftDown(0);
        return ret;
    }
    // 每次 去除 最后一个 和后面比较
    private void siftDown(int k) {
        while (leftChild(k) < data.getSize()) {
            int j = leftChild(k);
            if (j + 1 < data.getSize() && data.get(j+1).compareTo(data.get(j))>0) {
                j = rightChild(k);
            }

            if(data.get(k).compareTo(data.get(j))>0)
                break;

            data.swap(k,j);
            k = j;
        }
    }

    // 去出最大一个，并且 替换成元素e

    public E replace(E e){
        E ret = data.get(0);
        data.set(0,e);
        shiftUp(0);
        return  ret;
    }

    // heapify o(n)
    public MaxHeap (E [] arr){
        data = new Array<>(arr);
        for(int i = parent(arr.length-1);i>=0;i--){
            siftDown(i);
        }
    }




}

File: python/heap/max_heap.py
from Array import Array

class MaxHeap:
    def __init__(self, arr =None,capacity = None):
        if isinstance(arr, Array):
            self._data = arr
            for i in range(self._parent(arr.get_size() - 1), -1, -1):
                self._sift_down(i)
            return

        if not capacity:
            self._data = Array()
        else:
            self._data = Array(capacity=capacity)

    def size(self):
        return self._data.get_size()

    def is_empty(self):
        return self._data.is_empty()

    def _parent(self, index):
        if index == 0:
            raise ValueError('index-0 doesn\'t have parent.')
        return (index - 1) // 2

    # 都是简答的数组
    def _left_child(self, index):
        return index * 2 + 1
    
    def _right_child(self, index):
        return index * 2 + 2
    
    # 添加到最后的
    def add(self, e):
        self._data.add_last(e)
        self._sift_up(self._data.get_size() - 1)
    

    def _sift_up(self, k):
        while k > 0 and self._data.get(k) > self._data.get(self._parent(k)):
            self._data.swap(k, self._parent(k))
            k = self._parent(k)

    def _sift_down(self,k):
        while self._left_child(k) < self._data.get_size():
            j = self._left_child(k)
            if j + 1 < self._data.get_size() and self._data.get(j + 1) > self._data.get(j):
                j = self._right_child(k)
                if self._data.get(k) > self._data.get(j):
                    break
                self._data.swap(k,j)
                k = j

            
    def find_max(self):
        if self._data.get_size() == 0:
            raise ValueError('Can not find_max when heap is empty.')
        return self._data.get(0)

    def replace(self,e):
        ret = self.find_max()
        self._data.set(0, e)
        self._sift_down(0)
        return ret
    
    def extract_max(self):
        ref = self.find_max()
        self._data.swap(0,self._data.get_size()-1)
        self._data.remove_last()
        self._sift_down(0)
        return ref

    def HeapSort(self,alist):
        MaxHeap(alist)
        sortedList=[self.extract_max() for x in range(self._data.get_size())]
        sortedList.reverse()
        return sortedList



if __name__ == '__main__':
    n = 10
    from time import time

    start_time1 = time()
    max_heap = MaxHeap()
    from random import randint
    for i in range(n):
        max_heap.add(randint(0, 1000))
 
    start_time2 = time()
    arr = Array()
    from random import randint
    for i in range(n):
        num = randint(0, 10)
        arr.add_last(num)
        
    max_heap = MaxHeap(arr)
    print(max_heap.extract_max())

    

